main/lsp/ExtractVariable.cc

```diff
@@ -268,4 +289,216 @@ vector<unique_ptr<TextDocumentEdit>> VariableExtractor::getEdits(LSPTypecheckerD
     res.emplace_back(move(docEdit));
     return res;
 }
+
+// This tree walk takes a ExpressionPtr and looks for nodes that are the same as that node
+class ExpressionPtrSearchWalk {
+    ast::ExpressionPtr *targetNode;
+    vector<const ast::ExpressionPtr *> enclosingScopeStack;
+    std::vector<core::LocOffsets> skippedLocs;
+
+    // NOTE: Might want to profile and switch to UnorderedSet.
+    bool shouldSkipLoc(core::LocOffsets loc) {
+        return absl::c_find_if(skippedLocs, [loc](auto l) { return l.contains(loc); }) != skippedLocs.end();
+    }
+
+public:
+    vector<pair<vector<const ast::ExpressionPtr *>, const ast::ExpressionPtr *>> matches;
+    ExpressionPtrSearchWalk(ast::ExpressionPtr *matchingNode, std::vector<core::LocOffsets> skippedLocs)
+        : targetNode(matchingNode), skippedLocs(skippedLocs) {}
+
+    void preTransformExpressionPtr(core::Context ctx, const ast::ExpressionPtr &tree) {
+        if (!tree.loc().exists()) {
+            return;
+        }
+
+        if (shouldSkipLoc(tree.loc())) {
+            return;
+        }
+
+        // This is probably going to be slow and we'll probably need to come up with some tricks
+        // to make it faster. On the other hand, because this walk is scoped to only the enclosing
+        // method, it might not be too bad.
+        // TODO: think about whether tree.deepEqual(targetNode) would be faster
+        // TODO: see how slow a pathological case is (a long chain of the same nodes with just the deepest
+        // node being different) Ex.
+        //   a(a(a(a(a(a(b)))))).deepEqual(a(a(a(a(a(a(a(a(a(a(a(b))))))))))))
+        if (targetNode->deepEqual(tree)) {
+            matches.push_back(pair(enclosingScopeStack, &tree));
+        }
+    }
+    void preTransformInsSeq(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.push_back(&tree);
+    }
+
+    void postTransformInsSeq(core::Context ctx, const ast::ExpressionPtr &tree) {
+        enclosingScopeStack.pop_back();
+    }
```

Comment:
Why are there cases here that aren't in the cases below, which only handle `If`, `Rescue`, and `While`?